Add support for usage without Home Manager (nixosModule)

errnoh opened the pull request:
NOTE: I would consider this a draft for now, though the module is working fine.

This PR introduces `nixosModule` for schizofox

Things to note:
- Currently only supports installation for a single user. This would be quite simple to change to a list of users, but they would all still have same configuration options so decided to go with this for now.
- The users home directory must be at `/home/<user>`, this could be checked with terminal commands if one wants to improve things.
- This introduces one additional (mandatory) option, only for nixosModule: `userName`
- Code duplicates almost all of homeManagerModule code, with just few changes to the parts that needed to be adjusted for it to work without Home Manager (and some lines modified by linter). That said, it does re-use all other parts of the module as the parts that require HM were nicely isolated from everything else.
- Reading themes from Home Manager GTK options has been disabled for nixosModule.
- Didn't really test searx-randomizer. It'll probably work fine, but haven't tested it.

## Pull Request type

<!-- Please try to limit your pull request to one type; submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [x] Documentation content changes
- [x] Other (please describe): Latest commit also refactors code from home-manager module in order to avoid duplicated code.

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A

## What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

- nixosModule that can be used in configurations that don't use Home Manager

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this does introduce a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

<!-- Any other information that is important to this PR, such as screenshots of how the component looks before and after the change. -->

Questions:
- There are probably a lot of ways to remove duplicate code that should be done before merging?
- Are maintainers ok with a module that only supports installation for single user? I'd say that very high number of NixOS installations are like that, so personally I'd consider it bringing value to the user base and thus worth it.

NotAShelf commented:
hank you for looking into this! I have been meaning to, but couldn't quite find the time...

A "hint" before you invest more time into this is that we can perhaps expose the base configuration as a package and pass options directly to `wrapFirefox` wrapper, which would allow re-using module options without having to duplicate them.

errnoh commented:
No problem, I'm in process of rewriting my nixos configurations and getting rid of Home Manager while at it :)

Now that I look at it, the code duplication now comes mostly from just the file content definitions. If those (and searx randomizer systemd unit) would be in some shared file it would actually be quite clean. I'll probably try that out at some point to see how it would look. I have a feeling that it might be quite clean result with surprisingly small amount of work.

It would cause a bit larger PR though as it would refactor both modules and move the shared parts to different location from the homemanager module directory in which they are right now.

errnoh commented:
@NotAShelf I guess "at some point" was not that far away :sweat_smile: 

The PR is much more intrusive now as it moves the files around, but code has only been changed in `flake/modules/default.nix` and `flake/modules/home-manager.nix`. The home manager logic hasn't been touched at all, only changes are either from the linter or from using variables defined in `common`.

So take a look and if possible check if the homeManagerModule version works as intended. Let me know what you think :)

NotAShelf commented on flake/modules/nixos/default.nix:
```diff
@@ -0,0 +1,122 @@
+    ];
+
+    # Use activation script to create symlinks
+    system.userActivationScripts.schizofox.text = ''
+      # Create target directories if they don't exist
+      for dir in "${firefoxConfigPath}" "${defaultProfile}/chrome"; do
+        mkdir -p "$dir"
+      done
+
+      ln -sfn "${firefoxConfigDerivation}/profiles.ini" "${firefoxConfigPath}/profiles.ini"
+      ln -sfn "${firefoxConfigDerivation}/userChrome.css" "${defaultProfile}/chrome/userChrome.css"
+      ln -sfn "${firefoxConfigDerivation}/userContent.css" "${defaultProfile}/chrome/userContent.css"
+      ln -sfn "${firefoxConfigDerivation}/user.js" "${defaultProfile}/user.js"
+      ${pkgs.nix}/bin/nix-store --add-root "${firefoxConfigPath}/gcroot" --realise "${firefoxConfigDerivation}"
+    '';
```
I do not like the reliance on activation scripts. They're something that NixOS has been trying to get  rid of for a while, and writing to user's `$HOME` with one is incredibly impure. Not to mention that it does not allow multi-user support in the future.

Supporting NixOS should be done **explicitly** with `wrapFirefox`. Firefox, as far as I know, can read policies from `/etc/mozilla`, which is in the spirit of a NixOS module

NotAShelf commented on flake/modules/nixos/option.nix:
```diff
@@ -0,0 +1,19 @@
+{
+  pkgs,
+  lib,
+  ...
+}:
+let
+  inherit (lib.options) mkOption;
+  inherit (lib.types) str;
+in
+{
+  options.programs.schizofox = {
+    userName = mkOption {
+      type = str;
+      default = "";
+      description = "Name of the user to install the configuration for";
+      example = "ryan";
+    };
+  };
+}
```
Schizofox has no place asking for a username. Please refer to my previous comment.

```suggestion
```

NotAShelf reviewed (changes requested):
Sorry for the delay in the review, I got off work very late yesterday.

First of all thank you for looking into this, Schizofox has been on life support because of my lack of free time. I think this PR is on a good path, but the current implementation is *not it*.

1. Activation scripts are, simply put, bad.
2. Asking for a username partially replaces HM, but not enough to be effective

I would be much more in favor of using `wrapFirefox` for this case, you can see how the NixOS module handles things [here](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/programs/firefox.nix)

We may even hit two birds with one stone here, and move the Schizofox core to a package exposed by the flake, which takes a `cfg` attribute that NixOS and Home-Manager modules populate. That might also allow standalone support without relying on platform-specific solutions.

In short, I'm going to have to ask you to revise the current approach. `userActivationScripts` is a no-go, and I see more potential in using less-platform specific options. Moving modules to a "common" module is in the good direction, let's see if we can use it better.

errnoh commented:
Yup, no need for me to try to defend those, exactly the reasons I prefixed this with "consider this a draft" :)

The activationscript part is one of those "when you have a hammer, everything looks like a nail" situations. When investigating options to replace HM with there are basically two ways to implement such functionality: either you do it through activationscripts, or with systemd modules. The former are simpler code-wise and supported on each platform, while the latter are less of a hack :) Moving from activationscript to user-specfiic systemd unit would be quite simple and move it towards better direction, but as you said you want to avoid the unidiomatic approach of setting user names on schizofox side so that would be wrong approach as well.

So I think you're right here and a lot of this could be done with wrapFirefox directly and it would stay idiomatic. From couple minutes of browsing my initial feeling is that `profiles.ini` would be irrelevant as the whole wrapped schizofox would use relevant policies anways, and the data that's currently in `user.js` is the one that would be used to generate said policies in wrapFirefox. 

That leaves `userChrome.css` and `userContent.css`. Apparently there's a way to deal with them without having to mess with home directories either. There's a config option `toolkit.legacyUserProfileCustomizations.stylesheets = true` that seems to allow globally configured css in `/usr/lib/firefox/browser/chrome/`, that might be the route to do this without any modifications in user home directory.

errnoh commented:
Did some work on it. Still draft mode, could use cleanup here and there and the css doesn't work yet (tested usr/lib/ and lib/ both), but leaving current version here to keep discussion going

errnoh commented:
Found a working method to write the css files. It's a bit sketchy but seems like a common enough pattern: the global config file supports autoconfig that can also write to profile directory when the browser starts. Examples of that method being user
[here](https://github.com/NixOS/nixpkgs/blob/3efc5280fe028ab010747dc81d476544dbb87bec/pkgs/applications/networking/browsers/firefox/mobile-config.nix#L25) and [here](https://github.com/adriankarlen/textfox/blob/e0d002eea7a5a3791bbf3f4b64c9197a531a4f96/nix/pkgs/wrapTextfox.nix#L48).

Basically ensuring that each time new profile is created the CSS files are also written to the profile directory.

errnoh commented:
@NotAShelf when you happen to have time this might be a good stop to do a round of feedback on the approach taken here.

There are some interesting outcomes from this. For example when using the nixosModule schizofox basically defines the default configuration, but everything is overridable on profile level (unless we want to lock some/all of configured options). This also removes the need to define default profile as we don't need to know the profile directory on install time. Another benefit is that one can just spawn as many new profiles of schizofox they want with good defaults, customize each one and use those for different purposes fine grained control of config options for each.

For Home Manager configuration the logic should be same as previously for now. Afaik that does mean that creating a new profile does not inherit any of the configuration options (only the policies) and the configurations likely aren't similarly overridable from browser as the user.js is symlinked from nix store.

NotAShelf reviewed (changes requested):
Okay, it looks like two years of technical debt has finally caught up to us. While reviewing I've noticed *how much I don't like the current implementation*, but that is not necessarily relevant to this PR. Mozilla really has not made it very easy to manage Firefox declarative. 'tis a shame.

I've requested to changes, there are more diffs of such nature but I would like to emphasize that our official formatter is **Alejandra**. Nixfmt advertises the ugliest diffs I know off, and I would appreciate if you could undo the formatting relateed changes so that it becomes slightly easier for me to review in the future.

In terms of the implementation, it looks fine. Given how annoying it is to work with Mozilla's directory structure, I am not entirely sure if we can do better.

You already know my thoughts about using `wrapFirefox` on things, so I'm not going to repeat those. Keep up the good work!

errnoh commented on flake/modules/common/firefox/default.nix:
```diff
@@ -20,14 +21,17 @@
-      else "schizofox %U";
+    exec = if cfg.security.wrapWithProxychains then "proxychains4 schizofox %U" else "schizofox %U";
     icon = "${logo}";
     terminal = false;
-    categories = ["Network" "WebBrowser"];
-    mimeTypes = ["text/html" "text/xml"];
+    categories = [
+      "Network"
+      "WebBrowser"
+    ];
+    mimeTypes = [
+      "text/html"
+      "text/xml"
+    ];
```
Yeah, should've changed the formatter as soon as I noticed it introducing bunch of formatting related changes. It wasn't that visible initially when the PR barely touched any non-nixosmodule files but now it's making it much harder to read the diffs. (and even then, it's not like that would be acceptable anyways as the project does specify the formatter in its flake. It might be quite annoying process to try to get neovim to use formatter defined in projects flake specifically, but I wonder if there would be some clean way to do a local git hook of "if flake has formatter configured run that to modified files before commit")

errnoh commented:
Should be more readable now. Reverted the changes caused by different formatter and changed the comment on the autoconfig script to better reflect what it does.

kittyandrew commented:
Very excited about work in this direction, since home-manager module has been a pain needlessly it seems.
Did this become stale?

kittyandrew commented:
> I'm using the branch as my daily driver, but the PR has indeed become quite stale.

thanks for clarification, I might start using it in the meantime as well, knowing that it was left in the working state

> If you can merge the conflicts with main, I can merge this as is for those interested in using the feature.

sounds reasonable, @errnoh is there any chance of you resolving the conflicts for this?

errnoh commented:
Yup, should be simple enough to fix once I get back home :+1:

errnoh commented:
Hmm, one small question. This part feels weird: https://github.com/schizofox/schizofox/pull/130/files#diff-a8d1358476361984a5dbdb8af907e9b8a2c4d6569b1a5e9ec6b9991dd964c5a0R179-R186

Can't we just have `(sloth.concat' sloth.homeDir "/.mozilla")` and then `(sloth.concat' sloth.homeDir cfg.misc.customMozillaFolder.path)` if `cfg.misc.customMozillaFolder.enable` is toggled?

The else-clause doesn't seem necessary as it previously was just `(sloth.concat [sloth.homeDir "/.mozilla"])`

errnoh commented:
For the record, I made this change that reflects the thoughts in the previous message:
https://github.com/schizofox/schizofox/pull/129/files#diff-8756ba5db7573cfeda226de0cb210a38341bc6fa224acdf33969d2ff63b29186R218-R220

NotAShelf commented:
No, I'm still at work. My guess is that the problem is somewhere in common, as it is an eval failure directly on rebuild.

errnoh commented:
Found the problem, I'll post a PR to fix it.

errnoh commented:
Btw I did test this version (with minimal config options though) and it does fix the issue of module not working, there's at least one ugly thing that should be sorted out at some point: seems like the user.js logic that creates css files for nixosmodule isn't valid for the user.js in users home directory. This causes parse errors to be outputted to logs when browser is started.

I checked the browser itself and the options are being parsed and applied properly from the same file, so it should not affect user experience, but it does leave ugly logs.

kittyandrew commented:
very nice guys, thanks for your effort <3

kittyandrew commented:
(I'm using nixos module now)

I might be misunderstanding something, but looks like default extensions configurations are acting weird now.
First, I didn't get any of the default extensions, until I explicitly set `enableDefaultExtensions = true`, which wasn't a thing before.
Additionally, I cannot seem to enable darkreader (`darkreader.enable = true` is still there but has no effect now?)

errnoh commented:
While I haven't tested the nixosmodule version extensively with a lot of options, apart from refactoring I did try to keep the amount of changes to common parts to minimal. Based on the git history `enableDefaultExtensions` does seem to have been added 10 months ago so at least that shouldn't have changed in this commit.

I'm not familiar with darkreader part though. `programs.schizofox.extensions.darkreader.enable = true` does not do what's expected?

kittyandrew commented:
> I'm not familiar with darkreader part though. `programs.schizofox.extensions.darkreader.enable = true` does not do what's expected?

Yes, supposed to install a darkreader extension when enabled, but that didn't happen.

errnoh commented:
Hmm, I'm unable to replicate that issue. `programs.schizofox.extensions.darkreader.enable = true` when using the nixosmodule correctly installed darkreader and the extension seemed to work fine.
